kubectl-rook-ceph.sh

```diff
@@ -273,6 +274,91 @@ function run_purge_osd() {
     rook ceph osd remove --osd-ids=$1 --force-osd-removal=$force_removal"
 }
 
+function cluster_health() {
+  end_of_command_parsing "$@" # end of command tree
+
+  check_mon_pods_nodes
+  echo
+  check_mon_quorum
+  echo
+  check_osd_pod_count_and_nodes
+  echo
+  check_all_pods_status
+  echo
+  check_pg_are_active_clean
+  echo
+  check_mgr_pods_status_and_count
+}
+
+function check_mon_pods_nodes() {
+  echo -e "${INFO}Checking if atleast three mon pods are running on different nodes"
+  mon_unique_node_count=$(KUBECTL_NS_CLUSTER get pod -o=custom-columns=NAME:.metadata.name,NODE:.spec.nodeName | grep mon | grep -v canary | awk '{print $2}' | sort | uniq | wc -l)
+  if [ "$mon_unique_node_count" -lt 3 ]; then
+    echo -e "${WARNING}Atleast three mon pods should running on different node"
+  fi
+  KUBECTL_NS_CLUSTER get pod -o=custom-columns=NAME:.metadata.name,STATUS:.status.phase,NODE:.spec.nodeName | grep mon | grep --invert-match canary
+}
+
+function check_mon_quorum() {
+  echo -e "${INFO}Checking mon quorum and ceph health details"
+  ceph_health_details=$(KUBECTL_NS_OPERATOR exec deploy/rook-ceph-operator -- ceph health detail --conf="$CEPH_CONF_PATH")
+  if [[ "$ceph_health_details" = "HEALTH_OK" ]]; then
+    echo -e "$ceph_health_details"
+  elif [[ "$ceph_health_details" =~ "HEALTH_WARN" ]]; then
+    echo -e "${WARNING}$ceph_health_details"
+  elif [[ "$ceph_health_details" =~ "HEALTH_ERR" ]]; then
+    echo -e "${ERROR}$ceph_health_details"
+  fi
+}
+
+function check_osd_pod_count_and_nodes() {
+  echo -e "${INFO}Checking if atleast three osd pods are running on different nodes"
+  osd_unique_node_count=$(KUBECTL_NS_CLUSTER get pod -o=custom-columns=NAME:.metadata.name,NODE:.spec.nodeName | grep osd | grep -v prepare | awk '{print $2}' | sort | uniq | wc -l)
+  if [ "$osd_unique_node_count" -lt 3 ]; then
+    echo -e "${WARNING}Atleast three osd pods should running on different node"
+  fi
+  KUBECTL_NS_CLUSTER get pod -o=custom-columns=NAME:.metadata.name,STATUS:.status.phase,NODE:.spec.nodeName | grep osd | grep --invert-match prepare
+}
+
+function check_all_pods_status() {
+  echo -e "${INFO}Pods that are in 'Running' status"
+  KUBECTL_NS_OPERATOR get pod --field-selector status.phase=Running
+  if [ "$ROOK_OPERATOR_NAMESPACE" != "$ROOK_CLUSTER_NAMESPACE" ]; then
+    KUBECTL_NS_CLUSTER get pod --field-selector status.phase=Running
+  fi
+
+  echo
+  echo -e "${WARNING}Pods that are 'Not' in 'Running' status"
+  set +e # if there are no pod that are not running, then do no exit
```

Comment:
Perhaps we `set +e` for all commands that are printing status (as part of `cluster_health()` function above)? If a cluster is unhealthy, we want to print as much status as possible even if some commands fail.

Reply:
`set +e` should set it for the whole script as well. We can do `set +e` in the `run_cluster_health()` function and then `set -e` at the end to turn off error-ing for all helper funcions.

Reply:
done